### ISSUE TITLE
AnimationAction wrapping on time === duration

### ISSUE THREAD

**Issue by alexpineda** [mrdoob/three.js#24609]
**Describe the bug**

When using AnimationAction and setting the time = duration it loops around. Just wondering if this should be changed to test > than rather than >= than. 

**To Reproduce**
See code below.

***Code***

```js
animationMixer.setTime(clip.getDuration());
```

**Expected behavior**

It should not loop over until once it is passed duration.

**Comment by Mugen87** [comment]
Do you mind demonstrating the issue with a small live example? Maybe based on [misc_animation_keys](https://threejs.org/examples/misc_animation_keys)?

**Comment by alexpineda** [comment]
Yes I will provide an example.

**Comment by alexpineda** [comment]
Ok, so I made two examples, one using `misc_animation_keys` and the second using one of my gltf files.

The `misc_animation_keys` version seems to have no issue.
https://codepen.io/alexpineda/pen/eYrvrbg

My `gltf viewer` version seems to repeat frame 9 and 0. I still don't know why.
https://codepen.io/alexpineda/pen/eYrvKge


glb test file: https://drive.google.com/file/d/14JmKNSzKW2t0m0jQNbFABD-ruiGM9QfY/view?usp=sharing

**Comment by Suprhimp** [comment]
If we solve second example, can we change with the none equal version?

**Comment by AlaricBaraou** [comment]
@Suprhimp For the second example it's just an issue with your loop logic.

times[0] and times[times.length - 1] which are your first and last frame of animation aren't repeated, they are repeated in your example because you display them both manually separatly.

To illustrate look at this representation of 9 frames on a loop.
![9-frames](https://github.com/mrdoob/three.js/assets/7174039/a93e8778-cdd3-4d66-a38f-34587239e62a)

Your frame 1 and 9 should virtually happen at the same time. 

This is also why when using AnimationAction and setting the time to duration it loops around. 
I'm not sure how to explain why honestly, but to me it's just like how counting in modulo arithmetic resets to 0 after reaching a certain number.

@Mugen87 in my opinion there is no issue and this can be closed.

**Comment by cjones1313** [comment]
This is a big issue for me as well. I'm animating a morph target of positions where 0 is no deformation and duration is the full deformation. It works fine with pingpong because of the if statement below. For loop repeat it doesn't allow showing the full deformation. 

if ( pingPong && ( loopCount & 1 ) === 1 ) {
    // invert time for the "pong round"
    return duration - time;
}

I also use setTime to allow viewing specific frames within the animation when it's paused. This side effect becomes obvious to viewers in that scenario when they go to the last frame. 

Do you have a suggested work around other than subtracting a very small fraction if time is equal to the duration?

**Comment by AlaricBaraou** [comment]
I'm not sure to understand the use case where the current behaviour is an issue.
Can you share some code / demo and context ?
Maybe I can try suggesting something.

**Comment by cjones1313** [comment]
Thanks for the quick response. I added animation to the morphtargets example - https://threejs.org/examples/?q=mor#webgpu_morphtargets in codepen. 

https://codepen.io/crjones13/pen/eYXLOzJ

Notice at the end of the slider the (almost) sphere turns back to a cube. The desired behavior is a sphere (morphtargetinfluence of 1).

**Comment by AlaricBaraou** [comment]
I understand what you mean but if you're using a loop there is nothing that can be done. 

Let's even consider that a morphtarget influence of 1 doesn't automatically get converted back to 0 when you're using a loop and that instead we wait for strictly superior to 1 to loop back.
The sum of delta in your loop animation is more than likely never ever equal to 1. So that "frame" that you want to see, you're already seeing it interpolated between 9 and 10 in your example.

Unless we're talking about animating sprites, in which case you want to display every single one of them and there is no interpolation. 
But this is not the case here.

Does it make sense ?

**Comment by cjones1313** [comment]
I agree, while animating in a loop it's difficult to notice outside of a few edge cases that are minor. 

I'm doing data visualization where the deformation shown with morph targets is based on generated data. Going directly to the end time while animation is stopped to see the full deformation is a common use case. This is my main issue. 

Changing the code to "time > duration" as the original comment suggested would fix my issue. What is the downside to this change? Rather than counting modulo, resetting greater than duration is more appropriate in my opinion since times zero and duration are both unique times that should be represented.

**Comment by AlaricBaraou** [comment]
Here is a piece of code to explain the difference between "time > duration" and "time >= duration".

Here I simulate time incrementing by 1ms and interpolating between 0 and 1 as time progress.
The animation duration is 1 second.

```
let time = 0;
let animationTime = 0
const animationDuration = 1000;
const animationValues = [0, 1];

function interpolate(startValue, endValue, progress) {
    return startValue + (endValue - startValue) * progress;
}

for (time = 0; time <= 10 * animationDuration; time++) {
    
    
    if ( animationTime >= animationDuration ) { // change this line back to animationTime >= animationDuration to fix the offset
        animationTime -= animationDuration
    }
    // Adjust time for looping
    let progress = animationTime / animationDuration;
    const interpolatedValue = interpolate(animationValues[0], animationValues[1], progress);
    
    // Check if the interpolated value is approximately +/- 0.001 from 0 or 1
    if (interpolatedValue===0) {
        console.log(`Time when reach 0: ${time}ms`);
    }
    if (interpolatedValue===1) {
        console.log(`Time when reach 1: ${time}ms`);
    }

    animationTime++
}
```

If you ran the code above with your suggested change, you will notice that the loop starts 1ms late on each new iteration.

| Iteration | Time when reach 0 (ms) |  Time when reach 1 (ms) |
|-----------|------------------------|------------------------|
| 1         | 0                   | 1000                   |
| 2         |  never                   | 2000                   |
| 3         |  never               | 3000                   |
| 4         | never                   | 4000                   |
| 5         | never                   | 5000                   |
| 6         | never                   | 6000                   |
| 7         | never                   | 7000                   |
| 8         | never                   | 8000                   |
| 9         | never                   | 9000                   |

with the original >= you'd get 

| Iteration | Time when reach 0 (ms) |  Time when reach 1 (ms) |
|-----------|------------------------|------------------------|
| 1         | 0                   | never                   |
| 2         |  1000                   | never                   |
| 3         |  2000               | never                   |
| 4         | 3000                   | never                   |
| 5         | 4000                   | never                   |
| 6         | 5000                   | never                   |
| 7         | 6000                   | never                   |
| 8         | 7000                   | never                   |
| 9         | 8000                   | never                   |

This shows that you can only display one of the two state when looping, either the start or the end one. 

I'm not able to come up with a loop that can reach both 0 and 1 on each iteration without any side effect. 
To me the current behaviour make sense, but I'm open to be proven wrong.

I'm sorry I can't come up with clearer explaination

**Comment by cjones1313** [comment]
Thank you for the clear and helpful explanation. I don't see a solution to reach 0 and 1 yet either.

Of those two choices, I'm arguing that ">" is preferable because it allows manually setting the time (with mixer.setTime) to any time, including 0 and duration. While ">=" prevents setting the time to duration. And the difference while animating is minor.

I understand your preference though and have resolved my issues by wrapping setTime with a function that applies the small value subtraction work around.



Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
